Source/DafnyCore/Compilers/Compiler-java.cs

```diff
@@ -2361,41 +2361,50 @@ private class GenericArrayElementLvalue : ILvalue {
 
-      // Create the jar, in the specified output directory (dafnyProgramName is an absolute path to the desired output file)
-      System.IO.Directory.CreateDirectory(Path.GetDirectoryName(dafnyProgramName));
-      string jarPath = Path.GetFullPath(Path.ChangeExtension(dafnyProgramName, ".jar"));
-      var args = callToMain == null ?
-          new List<string> { "cf", jarPath }
-          : new List<string> { "cfe", jarPath, simpleProgramName };
-      var jarCreationProcess = PrepareProcessStartInfo("jar", args.Concat(classfiles));
-      jarCreationProcess.WorkingDirectory = Path.GetFullPath(Path.GetDirectoryName(targetFilename));
-      if (0 != RunProcess(jarCreationProcess, outputWriter, "Error while creating jar file: " + jarPath)) {
+      var simpleProgramName = Path.GetFileNameWithoutExtension(targetFilename);
+      var jarPath = Path.GetFullPath(Path.ChangeExtension(dafnyProgramName, ".jar"));
+      if (!createJar(callToMain == null ? null : simpleProgramName,
+                     jarPath,
+                     Path.GetFullPath(Path.GetDirectoryName(targetFilename)),
+                     classFiles,
+                     outputWriter)) {
         return false;
       }
 
       // Keep the build artifacts if --spill-translation is true
-      // But keep them for legacy CLI if the -spillTargetCode is 0, (because users and tests likely depend on it), but deprecate
-      if (DafnyOptions.O.UsingNewCli && DafnyOptions.O.SpillTargetCode == 0) {
-        System.IO.Directory.Delete(targetDirectory, true);
+      // But keep them for legacy CLI so as not to break old behavior
+      if (DafnyOptions.O.UsingNewCli) {
+        if (DafnyOptions.O.SpillTargetCode == 0) {
+          System.IO.Directory.Delete(targetDirectory, true);
+        } else {
+          classFiles.ForEach(f => System.IO.File.Delete(f));
+        }
       }
 
-      // When verbose, report the generated artifact
       if (DafnyOptions.O.CompileVerbose) {
         // For the sake of tests, just write out the filename and not the directory path
-        outputWriter.WriteLine("Wrote " + (callToMain != null ? "executable" : "library") + " jar " + Path.GetFileName(jarPath));
+        var fileKind = callToMain != null ? "executable" : "library";
+        outputWriter.WriteLine($"Wrote {0} jar {1}", fileKind, Path.GetFileName(jarPath));
       }
 
       return true;
     }
 
+
+    public bool createJar(string/*?*/ entryPointName, string jarPath, string rootDirectory, List<string> files, TextWriter outputWriter) {
+      System.IO.Directory.CreateDirectory(Path.GetDirectoryName(jarPath));
+      var args = entryPointName == null ? // If null, then no entry point is added
+          new List<string> { "cf", jarPath }
+          : new List<string> { "cfe", jarPath, entryPointName };
+      var jarCreationProcess = PrepareProcessStartInfo("jar", args.Concat(files));
+      jarCreationProcess.WorkingDirectory = rootDirectory;
```

Comment:
What is setting the working directory for? Can we not use fully qualified paths for the input files?

Reply:
I don't believe so. At least as I have always created jar files from the directory that would be the directory used in a CLASSPATH were it not a jar file, and the files within it are relative to that location so that their internal paths match the java package structure. Perhaps it works more generally, but I don't think so -- I've always done it this way.